File: hosts/gui/contract/src/session/transcript.rs
//! Transcript view-models: what a renderer draws.
//!
//! Mirrors `@veyyon/wire/presentation/transcript`. One block per message
//! variant the session can hold, carrying display-ready text and flags — never
//! provider payloads or tool argument objects.

use serde::{Deserialize, Serialize};

/// Stable identity of a block across updates. Assigned by the session, opaque
/// here.
pub type BlockId = String;

/// A file or image the operator attached to a message.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Attachment {
	pub kind:           AttachmentKind,
	/// Display name, already shortened for presentation.
	pub name:           String,
	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub byte_size:      Option<u64>,
	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub line_count:     Option<u64>,
	/// Why the content was not included. Absent when it was.
	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub omitted_reason: Option<OmittedReason>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum AttachmentKind {
	File,
	Image,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum OmittedReason {
	TooLarge,
	Binary,
	NotReplicated,
}

/// One span of an assistant turn, in emission order.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "kebab-case")]
pub enum AssistantSegment {
	Text {
		text: String,
	},
	Thinking {
		text:     String,
		redacted: bool,
	},
	#[serde(rename_all = "camelCase")]
	ToolCall {
		tool_call_id: String,
		tool_name:    String,
		input:        String,
	},
	#[serde(rename_all = "camelCase")]
	Image {
		mime_type: String,
		alt_text:  String,
	},
}

/// Lifecycle of a tool call as the renderer sees it.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ToolStatus {
	Pending,
	Running,
	Succeeded,
	Failed,
	Aborted,
	Rejected,
}

/// Why an assistant turn stopped, reduced to what a renderer displays.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum TurnStopReason {
	Complete,
	MaxTokens,
	ToolCall,
	Aborted,
	Error,
}

/// Token accounting for one assistant turn.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TurnUsage {
	pub input:       u64,
	pub output:      u64,
	pub cache_read:  u64,
	pub cache_write: u64,
	/// Reasoning tokens the provider billed separately, when it reported them.
	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub reasoning:   Option<u64>,
	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub cost_usd:    Option<f64>,
}

/// Presentation weight a host asked for.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum Level {
	Info,
	Warning,
	Error,
}

/// Every shape the transcript can hold.
///
/// Exhaustive over the TypeScript union. A renderer that matches on this
/// without a wildcard stops compiling when a member is added, which is the
/// point: a new message kind that silently renders as nothing is a message the
/// operator never sees.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "kebab-case")]
pub enum TranscriptBlock {
	#[serde(rename_all = "camelCase")]
	UserMessage {
		id:          BlockId,
		text:        String,
		attachments: Vec<Attachment>,
		timestamp:   i64,
	},
	/// A developer or system turn the operator can see: rules, injected
	/// instructions.
	#[serde(rename_all = "camelCase")]
	DeveloperMessage { id: BlockId, text: String, timestamp: i64 },
	#[serde(rename_all = "camelCase")]
	AssistantMessage {
		id:            BlockId,
		segments:      Vec<AssistantSegment>,
		/// Model identity as displayed.
		model:         String,
		stop_reason:   TurnStopReason,
		#[serde(default, skip_serializing_if = "Option::is_none")]
		usage:         Option<TurnUsage>,
		/// Set when the turn ended in a provider or transport failure.
		#[serde(default, skip_serializing_if = "Option::is_none")]
		error_message: Option<String>,
		/// True while the turn is still streaming.
		streaming:     bool,
		timestamp:     i64,
	},
	#[serde(rename_all = "camelCase")]
	ToolExecution {
		id:           BlockId,
		tool_call_id: String,
		tool_name:    String,
		status:       ToolStatus,
		/// Arguments rendered for display; secrets already redacted.
		input:        String,
		/// Result text rendered for display. Absent until the call finishes.
		#[serde(default, skip_serializing_if = "Option::is_none")]
		output:       Option<String>,
		#[serde(default, skip_serializing_if = "Option::is_none")]
		error:        Option<String>,
		#[serde(default, skip_serializing_if = "Option::is_none")]
		duration_ms:  Option<u64>,
		timestamp:    i64,
	},
	#[serde(rename_all = "camelCase")]
	BashExecution {
		id:        BlockId,
		command:   String,
		output:    String,
		/// `None` when the process was signalled rather than exiting.
		exit_code: Option<i32>,
		#[serde(default, skip_serializing_if = "Option::is_none")]
		signal:    Option<String>,
		cancelled: bool,
		timestamp: i64,
	},
	#[serde(rename_all = "camelCase")]
	PythonExecution {
		id:        BlockId,
		code:      String,
		output:    String,
		exit_code: Option<i32>,
		cancelled: bool,
		timestamp: i64,
	},
	/// A host-defined message with no meaning here beyond its text.
	#[serde(rename_all = "camelCase")]
	Custom {
		id:          BlockId,
		/// Discriminator the host assigned.
		custom_kind: String,
		text:        String,
		level:       Level,
		timestamp:   i64,
	},
	#[serde(rename_all = "camelCase")]
	Hook { id: BlockId, hook_name: String, text: String, timestamp: i64 },
	#[serde(rename_all = "camelCase")]
	BranchSummary {
		id:             BlockId,
		summary:        String,
		/// Messages the branch replaced.
		replaced_count: u64,
		timestamp:      i64,
	},
	#[serde(rename_all = "camelCase")]
	CompactionSummary {
		id:               BlockId,
		summary:          String,
		/// Messages compaction folded into the summary.
		replaced_count:   u64,
		/// Tokens the compaction reclaimed, when measured.
		#[serde(default, skip_serializing_if = "Option::is_none")]
		reclaimed_tokens: Option<u64>,
		timestamp:        i64,
	},
	#[serde(rename_all = "camelCase")]
	FileMention { id: BlockId, files: Vec<Attachment>, timestamp: i64 },
	/// A failure with no message of its own: a transport reset, a rejected
	/// request.
	#[serde(rename_all = "camelCase")]
	Error {
		id:          BlockId,
		message:     String,
		/// True when the session can continue; false when the turn is dead.
		recoverable: bool,
		timestamp:   i64,
	},
}

impl TranscriptBlock {
	/// The block's id, whichever variant it is. Every variant carries one, and
	/// the session addresses updates by it.
	pub fn id(&self) -> &BlockId {
		match self {
			TranscriptBlock::UserMessage { id, .. }
			| TranscriptBlock::DeveloperMessage { id, .. }
			| TranscriptBlock::AssistantMessage { id, .. }
			| TranscriptBlock::ToolExecution { id, .. }
			| TranscriptBlock::BashExecution { id, .. }
			| TranscriptBlock::PythonExecution { id, .. }
			| TranscriptBlock::Custom { id, .. }
			| TranscriptBlock::Hook { id, .. }
			| TranscriptBlock::BranchSummary { id, .. }
			| TranscriptBlock::CompactionSummary { id, .. }
			| TranscriptBlock::FileMention { id, .. }
			| TranscriptBlock::Error { id, .. } => id,
		}
	}

	/// When the block was produced, in milliseconds since the epoch.
	pub fn timestamp(&self) -> i64 {
		match self {
			TranscriptBlock::UserMessage { timestamp, .. }
			| TranscriptBlock::DeveloperMessage { timestamp, .. }
			| TranscriptBlock::AssistantMessage { timestamp, .. }
			| TranscriptBlock::ToolExecution { timestamp, .. }
			| TranscriptBlock::BashExecution { timestamp, .. }
			| TranscriptBlock::PythonExecution { timestamp, .. }
			| TranscriptBlock::Custom { timestamp, .. }
			| TranscriptBlock::Hook { timestamp, .. }
			| TranscriptBlock::BranchSummary { timestamp, .. }
			| TranscriptBlock::CompactionSummary { timestamp, .. }
			| TranscriptBlock::FileMention { timestamp, .. }
			| TranscriptBlock::Error { timestamp, .. } => *timestamp,
		}
	}
}

#[cfg(test)]
mod tests {
	use super::*;

	/// A block arrives as the session sends it: camelCase fields, kebab-case
	/// tag, absent optionals. This is the exact byte shape the TypeScript
	/// builder emits.
	#[test]
	fn a_block_deserializes_from_the_wire_shape() {
		let json = r#"{
			"kind": "tool-execution",
			"id": "b7",
			"toolCallId": "call_1",
			"toolName": "read",
			"status": "succeeded",
			"input": "{\"path\":\"src/main.rs\"}",
			"output": "fn main() {}",
			"durationMs": 12,
			"timestamp": 1730000000000
		}"#;
		let block: TranscriptBlock = serde_json::from_str(json).expect("deserializes");
		let TranscriptBlock::ToolExecution { tool_name, status, duration_ms, error, .. } = &block
		else {
			panic!("wrong variant: {block:?}");
		};
		assert_eq!(tool_name, "read");
		assert_eq!(*status, ToolStatus::Succeeded);
		assert_eq!(*duration_ms, Some(12));
		assert_eq!(*error, None);
		assert_eq!(block.id(), "b7");
		assert_eq!(block.timestamp(), 1_730_000_000_000);
	}

	/// An absent optional stays absent on the way back out. A renderer that
	/// re-emits a block must not turn `output?: string` into `output: null`,
	/// which the TypeScript side reads as a present empty result.
	#[test]
	fn an_absent_optional_is_not_serialized_as_null() {
		let block = TranscriptBlock::ToolExecution {
			id:           "b1".into(),
			tool_call_id: "c1".into(),
			tool_name:    "read".into(),
			status:       ToolStatus::Running,
			input:        "{}".into(),
			output:       None,
			error:        None,
			duration_ms:  None,
			timestamp:    0,
		};
		let json = serde_json::to_string(&block).expect("serializes");
		assert!(!json.contains("null"), "{json}");
		assert!(!json.contains("output"), "{json}");
		assert!(json.contains(r#""kind":"tool-execution""#), "{json}");
	}

	/// A signalled process has no exit code, and that null has to survive: it is
	/// the difference between "killed" and "exited 0".
	#[test]
	fn a_null_exit_code_survives_both_directions() {
		let json = r#"{
			"kind": "bash-execution",
			"id": "b2",
			"command": "sleep 1",
			"output": "",
			"exitCode": null,
			"signal": "SIGTERM",
			"cancelled": true,
			"timestamp": 0
		}"#;
		let block: TranscriptBlock = serde_json::from_str(json).expect("deserializes");
		let TranscriptBlock::BashExecution { exit_code, signal, .. } = &block else {
			panic!("wrong variant");
		};
		assert_eq!(*exit_code, None);
		assert_eq!(signal.as_deref(), Some("SIGTERM"));

		let out = serde_json::to_string(&block).expect("serializes");
		assert!(out.contains(r#""exitCode":null"#), "{out}");
	}

	/// Segments are tagged the same way blocks are, and their order is the
	/// emission order the model produced.
	#[test]
	fn assistant_segments_keep_their_order_and_tags() {
		let json = r#"{
			"kind": "assistant-message",
			"id": "a1",
			"segments": [
				{ "kind": "thinking", "text": "hm", "redacted": false },
				{ "kind": "text", "text": "here" },
				{ "kind": "tool-call", "toolCallId": "c1", "toolName": "read", "input": "{}" },
				{ "kind": "image", "mimeType": "image/png", "altText": "a chart" }
			],
			"model": "anthropic/claude-sonnet-4",
			"stopReason": "tool-call",
			"streaming": true,
			"timestamp": 0
		}"#;
		let block: TranscriptBlock = serde_json::from_str(json).expect("deserializes");
		let TranscriptBlock::AssistantMessage { segments, stop_reason, streaming, usage, .. } =
			&block
		else {
			panic!("wrong variant");
		};
		assert_eq!(segments.len(), 4);
		assert!(matches!(segments[0], AssistantSegment::Thinking { redacted: false, .. }));
		assert!(matches!(segments[1], AssistantSegment::Text { .. }));
		assert!(matches!(segments[2], AssistantSegment::ToolCall { .. }));
		assert!(matches!(segments[3], AssistantSegment::Image { .. }));
		assert_eq!(*stop_reason, TurnStopReason::ToolCall);
		assert!(*streaming);
		assert_eq!(*usage, None);
	}

	/// Every block variant round-trips through the wire shape unchanged. A
	/// variant whose field names or tag do not survive is one the GUI silently
	/// drops.
	#[test]
	fn every_block_variant_round_trips() {
		for block in crate::fixtures::transcript_blocks() {
			let json = serde_json::to_string(&block).expect("serializes");
			let back: TranscriptBlock =
				serde_json::from_str(&json).unwrap_or_else(|error| panic!("{json}: {error}"));
			assert_eq!(back, block, "{json}");
		}
	}
}

File: contracts/wire/src/presentation/transcript.ts
/**
 * Transcript view-models: what a renderer draws, with no dependency on the
 * agent runtime that produced it.
 *
 * One block per `AgentMessage` variant the session can hold, so a renderer that
 * handles every `kind` handles every message. Blocks carry rendered-ready text
 * and flags only — never provider payloads, tool argument objects, or anything
 * whose shape a renderer would have to know the agent to interpret.
 */

import type { ToolView } from "@veyyon/view";
import type { BranchSummaryView, CompactionSummaryView } from "./summary";

/** Stable identity of a block across updates. Assigned by the builder, opaque to the renderer. */
export type BlockId = string;

/** A file or image the operator attached to a message. */
export interface Attachment {
	kind: "file" | "image";
	/** Display name, already shortened for presentation. */
	name: string;
	/** Byte size when known. */
	byteSize?: number;
	/** Line count for a text file, when known. */
	lineCount?: number;
	/** Why the content was not included. Absent when it was. */
	omittedReason?: "too-large" | "binary" | "not-replicated";
	/** Raw base64 data for image payloads. */
	data?: string;
	/** MIME type for image payloads (e.g. image/png). */
	mimeType?: string;
	/** Original source URI or file link (e.g. file:///path/to/image.png). */
	uri?: string;
}

/** One span of an assistant turn, in emission order. */
export type AssistantSegment =
	| { kind: "text"; text: string }
	| { kind: "thinking"; text: string; redacted: boolean; rawThinking?: string }
	| { kind: "tool-call"; toolCallId: string; toolName: string; input?: string }
	| { kind: "image"; mimeType: string; altText: string }
	| { kind: "fallback" };

/** Lifecycle of a tool call as the renderer sees it. */
export type ToolStatus = "pending" | "running" | "succeeded" | "failed" | "aborted" | "rejected";

/** Why an assistant turn stopped, reduced to what a renderer displays. */
export type TurnStopReason = "complete" | "max-tokens" | "tool-call" | "aborted" | "error";

/** Token accounting for one assistant turn. */
export interface TurnUsage {
	input: number;
	output: number;
	cacheRead: number;
	cacheWrite: number;
	/** Reasoning tokens the provider billed separately, when it reported them. */
	reasoning?: number;
	costUsd?: number;
}

/** Display presentation of turn errors or retry recovery notes. */
export type AssistantErrorPresentation =
	| { kind: "none" }
	| { kind: "full"; text: string; isError: true }
	| { kind: "compact-recovered"; text: string; isError: false };

/** Display data shared by user and visible developer messages. */
export interface UserMessageView {
	text: string;
	synthetic?: boolean;
	imageLinks?: readonly (string | undefined)[];
}

/** Display data for an assistant turn. */
export interface AssistantMessageView {
	segments: readonly AssistantSegment[];
	model?: string;
	stopReason?: TurnStopReason;
	usage?: TurnUsage;
	/** Pre-resolved presentation of turn errors or retry recovery notes. */
	errorPresentation?: AssistantErrorPresentation;
	/** Exact reported thinking/reasoning token count for live streaming indicator. */
	reportedThinkingTokens?: number;
	timestamp?: number;
	provider?: string;
	responseId?: string;
}

export interface UserMessageBlock extends UserMessageView {
	kind: "user-message";
	id: BlockId;
	attachments: readonly Attachment[];
	timestamp: number;
}

/** A developer/system turn the operator can see (rules, injected instructions). */
export interface DeveloperMessageBlock extends UserMessageView {
	kind: "developer-message";
	id: BlockId;
	timestamp: number;
}

export interface AssistantMessageBlock extends AssistantMessageView {
	kind: "assistant-message";
	id: BlockId;
	/** Model identity as displayed, e.g. `"anthropic/claude-sonnet-4"`. */
	model: string;
	stopReason: TurnStopReason;
	/** True while the turn is still streaming. */
	streaming: boolean;
	timestamp: number;
}

export interface ToolExecutionImageItem {
	data?: string;
	mimeType?: string;
}

export interface ToolExecutionMultiFileItem {
	path: string;
	isError?: boolean;
	view?: ToolView;
	errorNotice?: string;
}

export interface ToolExecutionGenericDisplay {
	icon: "pending" | "running" | "done" | "error";
	argsPreview?: string;
	outputText?: string;
	isJson?: boolean;
}

export interface ToolExecutionPolicies {
	mergeCallAndResult?: boolean;
	callIsLiveWidget?: boolean;
	inline?: boolean;
	animatedPendingPreview?: boolean;
	animatedPartialResult?: boolean;
	forceFirstResultViewportRepaint?: boolean;
	forceResultViewportRepaintOnSettle?: boolean;
	backgroundTaskFrozen?: boolean;
	displaceable?: "job" | "todo";
	sealed?: boolean;
}

/** Display state of one entry in a grouped read card. */
export interface ReadEntryView {
	toolCallId: string;
	path: string;
	displayPaths?: string[];
	linkPath?: string;
	status: "pending" | "success" | "warning" | "notExecuted" | "error";
	correctedFrom?: string;
	contentText?: string;
	conflictCount?: number;
	codeStartLine?: number;
	codeLineNumbers?: Array<number | null>;
}

export interface ToolExecutionDisplay {
	toolLabel?: string;
	readEntry?: ReadEntryView;
	callView?: ToolView;
	resultView?: ToolView;
	multiFileViews?: readonly ToolExecutionMultiFileItem[];
	remainingPendingFiles?: number;
	notExecutedReason?: string;
	neverRan?: boolean;
	generic?: ToolExecutionGenericDisplay;
	images?: readonly ToolExecutionImageItem[];
	imageSourcePath?: string;
	policies?: ToolExecutionPolicies;
	failures?: Partial<
		Record<
			"call" | "result",
			{
				error: string;
				fallbackText?: string;
			}
		>
	>;
}

export interface ToolExecutionBlock {
	kind: "tool-execution";
	id: BlockId;
	toolCallId: string;
	toolName: string;
	status: ToolStatus;
	/** Arguments rendered for display; secrets already redacted by the builder. */
	input: string;
	/** Result text rendered for display. Absent until the call finishes. */
	output?: string;
	error?: string;
	/** Wall-clock duration in milliseconds once the call finished. */
	durationMs?: number;
	timestamp: number;
	/** Canonical neutral tool presentation state. */
	display?: ToolExecutionDisplay;
}

export interface BashExecutionBlock {
	kind: "bash-execution";
	id: BlockId;
	command: string;
	output: string;
	exitCode: number | null;
	signal?: string;
	cancelled: boolean;
	timestamp: number;
}

export interface PythonExecutionBlock {
	kind: "python-execution";
	id: BlockId;
	code: string;
	output: string;
	exitCode: number | null;
	cancelled: boolean;
	timestamp: number;
}

export interface AsyncResultJobDisplay {
	jobId?: string;
	type?: string;
	label?: string;
	durationMs?: number;
}

export interface AsyncResultCustomDisplay {
	variant: "async-result";
	jobs: readonly AsyncResultJobDisplay[];
}

export interface LateDiagnosticsFileDisplay {
	path?: string;
	summary?: string;
	errored?: boolean;
	messages?: readonly string[];
}

export interface LateDiagnosticsCustomDisplay {
	variant: "late-diagnostics";
	files: readonly LateDiagnosticsFileDisplay[];
}

export interface CollabPromptCustomDisplay {
	variant: "collab-prompt";
	from: string;
	text: string;
}

export interface SkillPromptCustomDisplay {
	variant: "skill-prompt";
	name: string;
	path?: string;
	args?: string;
	lineCount?: number;
	promptBytes?: number;
	text: string;
}

export interface IrcMessageCustomDisplay {
	variant: "irc";
	kind: "incoming" | "autoreply" | "relay";
	from?: string;
	to?: string;
	body?: string;
	replyTo?: string;
	timestamp?: number;
}

export interface AdvisorNoteDisplay {
	note: string;
	severity?: "nit" | "concern" | "blocker";
	advisor?: string;
}
export interface AdvisorCustomDisplay {
	variant: "advisor";
	notes: readonly AdvisorNoteDisplay[];
}

export interface BackgroundTanDispatchCustomDisplay {
	variant: "background-tan";
	jobId: string;
	work?: string;
	sessionFile?: string;
}

export interface HandoffSummaryCustomDisplay {
	variant: "handoff";
	summary: string;
}

export type CustomBlockDisplay =
	| AsyncResultCustomDisplay
	| LateDiagnosticsCustomDisplay
	| CollabPromptCustomDisplay
	| SkillPromptCustomDisplay
	| IrcMessageCustomDisplay
	| AdvisorCustomDisplay
	| BackgroundTanDispatchCustomDisplay
	| HandoffSummaryCustomDisplay;

/** A host-defined message with no runtime meaning to the renderer beyond its text. */
export interface CustomBlock {
	kind: "custom";
	id: BlockId;
	/** Discriminator the host assigned, e.g. `"notice"`. */
	customKind: string;
	text: string;
	/** Presentation weight the host asked for. */
	level: "info" | "warning" | "error";
	timestamp: number;
	display?: CustomBlockDisplay;
}

export interface HookBlock {
	kind: "hook";
	id: BlockId;
	hookName: string;
	text: string;
	timestamp: number;
	display?: CustomBlockDisplay;
	level?: "info" | "warning" | "error";
}

export interface BranchSummaryBlock extends BranchSummaryView {
	id: BlockId;
	timestamp: number;
}

export interface CompactionSummaryBlock extends CompactionSummaryView {
	id: BlockId;
	timestamp: number;
}

export interface FileMentionBlock {
	kind: "file-mention";
	id: BlockId;
	files: readonly Attachment[];
	timestamp: number;
}

/** A failure with no message of its own: a transport reset, a rejected request. */
export interface ErrorBlock {
	kind: "error";
	id: BlockId;
	message: string;
	/** True when the session can continue; false when the turn is dead. */
	recoverable: boolean;
	timestamp: number;
}

/**
 * Every shape the transcript can hold. Exhaustive over `AgentMessage`: a new
 * message variant is a new member here, and a renderer that switches on `kind`
 * without a default fails to compile until it handles the new one.
 */
export type TranscriptBlock =
	| UserMessageBlock
	| DeveloperMessageBlock
	| AssistantMessageBlock
	| ToolExecutionBlock
	| BashExecutionBlock
	| PythonExecutionBlock
	| CustomBlock
	| HookBlock
	| BranchSummaryBlock
	| CompactionSummaryBlock
	| FileMentionBlock
	| ErrorBlock;

/** Every `TranscriptBlock["kind"]`, as a value, so a sweep can enumerate the union at run time. */
export const TRANSCRIPT_BLOCK_KINDS = [
	"user-message",
	"developer-message",
	"assistant-message",
	"tool-execution",
	"bash-execution",
	"python-execution",
	"custom",
	"hook",
	"branch-summary",
	"compaction-summary",
	"file-mention",
	"error",
] as const satisfies readonly TranscriptBlock["kind"][];

/**
 * A new `TranscriptBlock` member that is missing from TRANSCRIPT_BLOCK_KINDS makes this fail to
 * compile, naming the member. `satisfies` above rejects a stale entry; this
 * rejects a missing one, so the table cannot drift from the union either way.
 */
type UnlistedTranscriptBlock = Exclude<TranscriptBlock["kind"], (typeof TRANSCRIPT_BLOCK_KINDS)[number]>;
const _transcript_block_kinds_is_exhaustive: UnlistedTranscriptBlock extends never ? true : UnlistedTranscriptBlock =
	true;
void _transcript_block_kinds_is_exhaustive;
